analytics.go

```diff
@@ -3,85 +3,123 @@ package analytics
 import (
 	"io"
 	"io/ioutil"
-	"sync"
 
 	"bytes"
 	"encoding/json"
 	"net/http"
 	"time"
-
-	"github.com/segmentio/backo-go"
-	"github.com/xtgo/uuid"
 )
 
 // Version of the client.
 const Version = "3.0.0"
 
-// Endpoint for the Segment API.
-const Endpoint = "https://api.segment.io"
-
-// Backoff policy.
-var Backo = backo.DefaultBacko()
-
-// Client which batches messages and flushes at the given Interval or
-// when the Size limit is exceeded. Set Verbose to true to enable
-// logging output.
-type Client struct {
-	Endpoint string
-	// Interval represents the duration at which messages are flushed. It may be
-	// configured only before any messages are enqueued.
-	Interval  time.Duration
-	Size      int
-	Logger    Logger
-	Verbose   bool
-	Transport http.RoundTripper
-	key       string
-	msgs      chan interface{}
-	quit      chan struct{}
-	shutdown  chan struct{}
-	uid       func() string
-	now       func() time.Time
-	once      sync.Once
+// This interface is the main API exposed by the analytics package.
+// Values that satsify this interface are returned by the client constructors
+// provided by the package and provide a way to send messages via the HTTP API.
+type Client interface {
+	io.Closer
+
+	// Queues a message to be sent by the client when the conditions for a batch
+	// upload are met.
+	// This is the main method you'll be using, a typical flow would look like
+	// this:
+	//
+	//	client := analytics.New(writeKey)
+	//	...
+	//	client.Enqueue(analytics.Track{ ... })
+	//	...
+	//	client.Close()
+	//
+	// The method returns an error if the message queue not be queued, which
+	// happens if the client was already closed at the time the method was
+	// called or if the message was malformed.
+	Enqueue(Message) error
 }
 
-// New client with write key.
-func New(key string) *Client {
-	c := &Client{
-		Endpoint:  Endpoint,
-		Interval:  5 * time.Second,
-		Size:      250,
-		Logger:    newDefaultLogger(),
-		Verbose:   false,
-		Transport: http.DefaultTransport,
-		key:       key,
-		msgs:      make(chan interface{}, 100),
-		quit:      make(chan struct{}),
-		shutdown:  make(chan struct{}),
-		now:       time.Now,
-		uid:       uid,
-	}
+type client struct {
+	Config
+	key string
+
+	// This channel is where the `Enqueue` method writes messages so they can be
+	// picked up and pushed by the backend goroutine taking care of applying the
+	// batching rules.
+	msgs chan interface{}
```

Comment:
Would it  be better to make this `chan message` rather than `interface{}`, and offload serialization to the grouting too?